src/Logger/Log.php

```diff
@@ -0,0 +1,338 @@
+<?php
+
+namespace Utopia\Logger;
+
+use Exception;
+use Utopia\Logger\Log\Breadcrumb;
+use Utopia\Logger\Log\User;
+
+class Log
+{
+    const TYPE_DEBUG = "debug";
+    const TYPE_ERROR = "error";
+    const TYPE_WARNING = "warning";
+    const TYPE_INFO = "info";
+    const TYPE_VERBOSE = "verbose";
+
+    const ENVIRONMENT_PRODUCTION = "production";
+    const ENVIRONMENT_STAGING = "staging";
+
+    /**
+     * @var float (required, set by default to microtime(true))
+     */
+    protected float $timestamp;
+
+    /**
+     * @var string (required, for example 'Log::TYPE_INFO')
+     */
+    protected string $type;
+
+    /**
+     * @var string (required)
+     */
+    protected string $message;
+
+    /**
+     * @var string (required)
+     */
+    protected string $version;
+
+    /**
+     * @var string (required)
+     */
+    protected string $environment;
+
+    /**
+     * @var string (required)
+     */
+    protected string $action;
+
+    /**
+     * @var array (optional)
+     */
+    protected array $tags = [];
+
+    /**
+     * @var array (optional)
+     */
+    protected array $extra = [];
+
+    /**
+     * @var string (optional)
+     */
+    protected string $namespace = "UNKNOWN";
+
+    /**
+     * @var string|null (optional)
+     */
+    protected ?string $server = null;
+
+    /**
+     * @var User|null (optional)
+     */
+    protected ?User $user = null;
+
+    /**
+     * @var Breadcrumb[] (optional)
+     */
+    protected array $breadcrumbs = [];
+
+    /**
+     * Log constructor.
+     */
+    public function __construct()
+    {
+        $this->timestamp = \microtime(true);
+    }
+
+    /**
+     * Set a type
+     *
+     * @param string $type (required, can be 'log', 'error' or 'warning')
+     * @return void
+     * @throws Exception
+     */
+    public function setType(string $type): void {
```

Comment:
```suggestion
    public function setType(string $type): void 
    {
```

I almost forgot. We need to have the opening brace in a new line according to PHP Fig standards. Please fix it wherever relevant

Reply:
Ill fix all manually, but can we make this part of the code-quality test? 🤔 Would be cool prevention for future.

Reply:
Only class and function blocks, or same for if/for/anything?

Reply:
^ Only func+class